### PR TITLE
Swift: upgrade to 5.9

### PR DESCRIPTION
**Pull request by AlexDenisov**

### CODE REVIEW

File: swift/schema.py
Comment (redsun82) [inline swift/schema.py:1246]: "may" makes it sound like the statement would be optional. Maybe
```suggestion
    An expression that wraps a statement which produces a single value.
```

File: swift/ql/test/extractor-tests/generated/expr/SingleValueStmtExpr/MISSING_SOURCE.txt
Comment (redsun82) [inline swift/ql/test/extractor-tests/generated/expr/SingleValueStmtExpr/MISSING_SOURCE.txt:1]: could you add a small `SingleValueStmtExpr` source snippet here and learn the tests? If not now, when? :wink:

File: swift/schema.py
Comment (AlexDenisov) [reply swift/schema.py:1246]: That's what the docs from compiler sources said 🤷

File: swift/ql/test/library-tests/dataflow/flowsources/CONSISTENCY/CfgConsistency.expected
Comment (geoffw0) [inline swift/ql/test/library-tests/dataflow/flowsources/CONSISTENCY/CfgConsistency.expected:5]: How concerned should we be about these?

**Review by geoffw0 — COMMENTED** [review]
The library changes and upgrade / downgrade scripts LGTM.

File: swift/ql/test/extractor-tests/generated/expr/SingleValueStmtExpr/MISSING_SOURCE.txt
Comment (AlexDenisov) [reply swift/ql/test/extractor-tests/generated/expr/SingleValueStmtExpr/MISSING_SOURCE.txt:1]: It's implicitly tested by the other tests: the reason I had to extract it as part of the upgrade is that it just eliminated huge subtrees in some PrintAST tests.
But I'll try to craft some small explicit example.

File: swift/ql/test/library-tests/dataflow/flowsources/CONSISTENCY/CfgConsistency.expected
Comment (AlexDenisov) [reply swift/ql/test/library-tests/dataflow/flowsources/CONSISTENCY/CfgConsistency.expected:5]: These are not the only ones we have, so I'm not sure if it gets much worse with these changes.

Comment (geoffw0) [reply swift/ql/test/library-tests/dataflow/flowsources/CONSISTENCY/CfgConsistency.expected:5]: ... I see that you've said elsewhere they should be addressed separately.  That's fine, as long as there's an issue and any breakage before then is limited.

File: swift/schema.py
Comment (redsun82) [reply swift/schema.py:1246]: Hmm, then we should probably double-check whether `Stmt` should not be `optional` (I would be surprised if it could though). If not, then I would drop the `may` regardless of the internal compiler docs (which by the way we shouldn't really copy :wink:).

File: swift/ql/test/library-tests/dataflow/flowsources/CONSISTENCY/CfgConsistency.expected
Comment (geoffw0) [reply swift/ql/test/library-tests/dataflow/flowsources/CONSISTENCY/CfgConsistency.expected:5]: > These are not the only ones we have, so I'm not sure if it gets much worse with these changes.

Fair enough.  My intuition is there's a decent chance these kinds of problems do cost us data flow and thus results, so they should not be ignored in the long run.

**Comment by geoffw0** [comment]
I'm happy, once @redsun82 and the checks and the DCA job are all happy.

File: swift/ql/test/extractor-tests/generated/expr/SingleValueStmtExpr/MISSING_SOURCE.txt
Comment (AlexDenisov) [reply swift/ql/test/extractor-tests/generated/expr/SingleValueStmtExpr/MISSING_SOURCE.txt:1]: Added [here](https://github.com/github/codeql/pull/14261/commits/df256e601af5ddc33d01b03f5b41164968ff9fcd)

File: swift/ql/test/library-tests/dataflow/flowsources/CONSISTENCY/CfgConsistency.expected
Comment (AlexDenisov) [reply swift/ql/test/library-tests/dataflow/flowsources/CONSISTENCY/CfgConsistency.expected:5]: I filed an issue. I think the fix should be trivial as we need to add some wires for the new AST node in CFG.

Comment (AlexDenisov) [reply swift/ql/test/library-tests/dataflow/flowsources/CONSISTENCY/CfgConsistency.expected:5]: And attempted to fix it here https://github.com/github/codeql/pull/14717

Comment (geoffw0) [reply swift/ql/test/library-tests/dataflow/flowsources/CONSISTENCY/CfgConsistency.expected:5]: 🎉

**Comment by geoffw0** [comment]
Well I don't understand what's going on with the change note, but I remain happy with this PR.  @redsun82 can approve if his concerns have been addressed.

> the QLDoc failure can be ignored as it will be fixed after https://github.com/github/codeql/pull/14715

I think when one of these two PRs is merged, the other will need updating.

**Comment by AlexDenisov** [comment]
> Well I don't understand what's going on with the change note, but I remain happy with this PR

AFAICT, currently we only have space for QL-related change notes. I'd like to include a note saying that we now support Swift 5.9, but previously I had to talk to a release manager to make this happen.

> I think when one of these two PRs is merged, the other will need updating.

For sure, I hope this one gets in first though 😅

File: swift/ql/test/extractor-tests/generated/expr/SingleValueStmtExpr/CONSISTENCY/CfgConsistency.expected
Comment (redsun82) [inline swift/ql/test/extractor-tests/generated/expr/SingleValueStmtExpr/CONSISTENCY/CfgConsistency.expected:2]: I think this means that the problem with the new dead ends is just that the new extracted AST node needs to be added to control flow as well. This can happen in a follow up PR though :+1:

Comment (AlexDenisov) [reply swift/ql/test/extractor-tests/generated/expr/SingleValueStmtExpr/CONSISTENCY/CfgConsistency.expected:2]: It's indeed in the followup PR https://github.com/github/codeql-c-team/issues/2042

**Comment by redsun82** [comment]
> unless we need to add that we now extract one more AST node

More or less this. From the library perspective, there is a new class, so it's probably worth mentioning that. ``* added `SingleValueStmtExpr`.`` would probably be enough, I think?

**Comment by AlexDenisov** [comment]
> > unless we need to add that we now extract one more AST node
> 
> More or less this. From the library perspective, there is a new class, so it's probably worth mentioning that. `` * added `SingleValueStmtExpr`. `` would probably be enough, I think?

I'll do in the followup PR then, I don't want to restart the whole CI process for this PR. (unless there is some other changes need to happen here).

**Comment by geoffw0** [comment]
> AFAICT, currently we only have space for QL-related change notes. I'd like to include a note saying that we now support Swift 5.9, but previously I had to talk to a release manager to make this happen.

I feel that the library change notes are an appropriate place to describe extractor changes (i.e. `swift\ql\lib\change-notes`), if we do not have separate extractor change notes.

**Comment by AlexDenisov** [comment]
@redsun82 @geoffw0 I added change note.

**Review by geoffw0 — APPROVED** [review]
I think this is ready to merge.  Please check that @redsun82 agrees before merging.

**Review by redsun82 — APPROVED** [review]
Thanks for the terrific work on this!

**Pull request merged**
